### PR TITLE
Add index_configs field to log bucket configuration

### PR DESCRIPTION
**Pull request by mudim**
Ability to manage log bucket index configurations by adding `index_configs` field to logging_bucket_config resources.

Addresses feature request raised in [hashicorp/terraform-provider-google#13287](https://github.com/hashicorp/terraform-provider-google/issues/13287)

```release-note:enhancement
logging: added `index_configs` field to `logging_bucket_config` resource
```

```release-note:enhancement
logging: added `index_configs` field to `logging_project_bucket_config` resource
```

### CODE REVIEW

**Comment by mudim** [comment]
Hi @rileykarson, following the previous comment from MM bot, as it's been more than 2 days, wanted to follow up on review regarding this PR. Appreciate your time. Thanks

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go [hunk 1]
```diff
@@ -88,6 +90,35 @@ See [Enabling CMEK for Logging Buckets](https://cloud.google.com/logging/docs/ro
...
+					Description: `The LogEntry field path to index.`,
+				},
+				"type": &schema.Schema{
+					Type:     schema.TypeString,
+					Required: true,
+					Description: `The type of data in this index
+Note that some paths are automatically indexed, and other paths are not eligible for indexing. See [indexing documentation]( https://cloud.google.com/logging/docs/view/advanced-queries#indexed-fields) for details.
+For example: jsonPayload.request.status`,
+				},
+				"create_time": &schema.Schema{
```
Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go:112]: We should leave this field off if not absolutely necessary `schema.TypeSet` does not play well with `Computed` fields, due to how diffs are calculated for sets, and Terraform tries to reuse old computed values.

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go [hunk 2]
```diff
@@ -405,3 +458,40 @@ func flattenCmekSettings(cmekSettings interface{}) []map[string]interface{} {
 
 	return []map[string]interface{}{data}
 }
+
+func expandIndexConfigs(originalIndexConfigs []interface{}) ([]map[string]interface{}, error) {
+	transformedIndexConfigs := make([]map[string]interface{}, 0, len(originalIndexConfigs))
+	for _, entry := range originalIndexConfigs {
+		original := entry.(map[string]interface{})
+		var validIndexType = false
```
Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go:466]: Can we just defer to the API error here? We're already in apply, let's not add extra complexity clientside.

If this was a ValidateFunc- which works at plan time- it may be worth keeping around.

File: mmv1/third_party/terraform/website/docs/r/logging_folder_bucket_config.html.markdown [hunk 3]
```diff
@@ -28,6 +28,22 @@ resource "google_logging_folder_bucket_config" "basic" {
 }
 ```
 
+Create logging bucket with index configs
+
+```hcl
```
Comment (rileykarson) [inline mmv1/third_party/terraform/website/docs/r/logging_folder_bucket_config.html.markdown:33]: Nit: spacing on these blocks. They should be 2-spaced spaces and are currently mixed tabs+spaces.

(All tabs would also work for consistency within the file, but the other entry in this file is also wrong)

File: mmv1/third_party/terraform/website/docs/r/logging_folder_bucket_config.html.markdown [hunk 4]
```diff
@@ -42,6 +58,19 @@ The following arguments are supported:
...
 * `retention_days` - (Optional) Logs will be retained by default for this amount of time, after which they will automatically be deleted. The minimum retention period is 1 day. If this value is set to zero at bucket creation time, the default time of 30 days will be used. Bucket retention can not be increased on buckets outside of projects.
 
+* `index_configs` - (Optional) A list of indexed fields and related configuration data. Structure is [documented below](#nested_index_configs).
+
+<a name="nested_index_configs"></a>The `index_configs` block supports:
+
+* `field_path` - The LogEntry field path to index.
+  Note that some paths are automatically indexed, and other paths are not eligible for indexing. See [indexing documentation]( https://cloud.google.com/logging/docs/view/advanced-queries#indexed-fields) for details.
+
+* `type` - The type of data in this index.
```
Comment (rileykarson) [inline mmv1/third_party/terraform/website/docs/r/logging_folder_bucket_config.html.markdown:68]: We may want to list enum values here!

**Comment by rileykarson** [comment]
Robot seems wrong here, ignore the breaking change detector / missing test messages from it.

**Comment by mudim** [comment]
Hi @rileykarson, thanks for taking time and review on it. Have addressed the comments raised above. 

One thing around the tests reported by MM as failures, ignoring the unrelated ones to this change, have looked at the 2 logging tests reported as failed. However, they pass locally both for ga and beta versions. Any feedback on that part would be helpful. 

Thanks

**Comment by rileykarson** [comment]
> TestAccLoggingBucketConfigBillingAccount_basic

This was a missing permission. It was in our manual bootstrapping guide but not applied to the billing account of our presubmits (which previously replayed logs from an old target).

> TestAccLoggingBucketConfigProject_locked

Failing at HEAD, filed https://github.com/hashicorp/terraform-provider-google/issues/16099

File: mmv1/third_party/terraform/website/docs/r/logging_organization_bucket_config.html.markdown [hunk 5]
```diff
@@ -27,6 +27,22 @@ resource "google_logging_organization_bucket_config" "basic" {
 }
 ```
 
+Create logging bucket with index configs
```
Comment (rileykarson) [inline mmv1/third_party/terraform/website/docs/r/logging_organization_bucket_config.html.markdown:30]: nit: Let's just roll the new field into the entry above.

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config_test.go [hunk 6]
```diff
@@ -523,3 +523,69 @@ func getLoggingBucketConfigs(context map[string]interface{}) map[string]string {
 	}
 
 }
+
+func TestAccLoggingBucketConfigProject_indexConfigs(t *testing.T) {
```
Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_bucket_config_test.go:527]: Can you add a test on one of the other targets? It's not a big deal if you can't run it locally, since I'll be able to in our CI.

**Review by rileykarson — COMMENTED** [review]
Sorry, forgot to click "Submit review" yesterday!

**Comment by mudim** [comment]
Thanks @rileykarson. Have pushed up changes to 
- Add a new test at org level
- Remove usage of `createTime` at resource read step
- Update markdown to collapse indexConfig into existing example

**Comment by mudim** [comment]
Hi @rileykarson, just wanted to follow up on previous changes addressing comments, when u get a chance pls. Thanks

**Comment by rileykarson** [comment]
Thanks! Running tests, sorry for the delay on my end.

**Comment by mudim** [comment]
> Thanks! Running tests, sorry for the delay on my end.

No problem, thanks again for taking time with this 👍

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go [hunk 7]
```diff
@@ -287,6 +314,13 @@ func resourceLoggingBucketConfigRead(d *schema.ResourceData, meta interface{}) e
 		return fmt.Errorf("Error setting cmek_settings: %s", err)
 	}
 
+	indexConfigs, ok := res["indexConfigs"]
+	if ok {
+		if err := d.Set("index_configs", flattenIndexConfigs(indexConfigs.([]interface{}))); err != nil {
```
Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go:319]: Ah- we want to set this all the time, or else we won't detect if the setting has been removed out of band. Terraform preserves the old value in state if you do nothing in Read.

You can safely remove the `ok`, the flattener will handle a `nil` appropriately.

File: mmv1/third_party/terraform/services/logging/resource_logging_project_bucket_config.go [hunk 8]
```diff
@@ -106,6 +106,35 @@ See [Enabling CMEK for Logging Buckets](https://cloud.google.com/logging/docs/ro
 			},
 		},
 	},
+	"index_configs": &schema.Schema{
```
Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_project_bucket_config.go:109]: This file needs `gofmt` run on it! `gofmt -w -s mmv1/third_party/terraform/services/logging/resource_logging_project_bucket_config.go` from the repo root should do it.

File: mmv1/third_party/terraform/services/logging/resource_logging_project_bucket_config.go [hunk 9]
```diff
@@ -106,6 +106,35 @@ See [Enabling CMEK for Logging Buckets](https://cloud.google.com/logging/docs/ro
...
+					Description: `The LogEntry field path to index.`,
+				},
+				"type": &schema.Schema{
+					Type:     schema.TypeString,
+					Required: true,
+					Description: `The type of data in this index
+Note that some paths are automatically indexed, and other paths are not eligible for indexing. See [indexing documentation]( https://cloud.google.com/logging/docs/view/advanced-queries#indexed-fields) for details.
+For example: jsonPayload.request.status`,
+				},
+				"create_time": &schema.Schema{
```
Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_project_bucket_config.go:128]: Ah- same as https://github.com/GoogleCloudPlatform/magic-modules/pull/9109#discussion_r1343168254, `Computed` values and `TypeSet` don't play nicely. We should remove this as well (and any associated expander/flattener code, but I think those are already updated since they were shared)

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config_test.go [hunk 10]
```diff
@@ -523,3 +523,129 @@ func getLoggingBucketConfigs(context map[string]interface{}) map[string]string {
...
+}
+
+resource "google_logging_organization_bucket_config" "basic" {
+	organization    = data.google_organization.default.organization
+	location  = "global"
+	retention_days = 30
+	description = "retention test 30 days"
+	bucket_id = "_Default"
+
+    index_configs {
```
Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_bucket_config_test.go:574]:
```suggestion
	index_configs {
```

(it's probably easier to just make this change in your editor)

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config_test.go [hunk 11]
```diff
@@ -523,3 +523,129 @@ func getLoggingBucketConfigs(context map[string]interface{}) map[string]string {
...
+				Config: testAccLoggingBucketConfigOrganization_indexConfigs(context, "INDEX_TYPE_STRING", "INDEX_TYPE_STRING"),
+			},
+			{
+				ResourceName:            "google_logging_organization_bucket_config.basic",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"organization"},
+			},
+			{
+				Config: testAccLoggingBucketConfigOrganization_indexConfigs(context, "INDEX_TYPE_STRING", "INDEX_TYPE_STRING"),
```
Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_bucket_config_test.go:549]: Can we change one of the values here, similar to TestAccLoggingBucketConfigProject_indexConfigs below?

File: mmv1/third_party/terraform/services/logging/resource_logging_project_bucket_config.go [hunk 12]
```diff
@@ -282,6 +316,13 @@ func resourceLoggingProjectBucketConfigRead(d *schema.ResourceData, meta interfa
 		return fmt.Errorf("Error setting cmek_settings: %s", err)
 	}
 
+	indexConfigs, ok := res["indexConfigs"]
+	if ok {
```
Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_project_bucket_config.go:320]: Same as above- we need to unconditionally write this so that we detect the removal of the value out-of-band.

**Review by rileykarson — COMMENTED** [review]
Acceptance test failures are all unrelated. The build-and-test failures are just a formatting issue that `gofmt` will fix. I noticed I missed something in Read and that `create_time` was still around in project while scanning through.

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config_test.go (hunk 10 above)
Comment (mudim) [reply mmv1/third_party/terraform/services/logging/resource_logging_bucket_config_test.go:574]: Odd that it's showing correct format locally, will check again 👍

**Comment by mudim** [comment]
Thanks @rileykarson . Have pushed up changes addressing recent comments 👍

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go [hunk 13]
```diff
@@ -287,6 +314,10 @@ func resourceLoggingBucketConfigRead(d *schema.ResourceData, meta interface{}) e
 		return fmt.Errorf("Error setting cmek_settings: %s", err)
 	}
 
+	if err := d.Set("index_configs", flattenIndexConfigs(indexConfigs.([]interface{}))); err != nil {
```
Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go:317]:
```suggestion
	if err := d.Set("index_configs", flattenIndexConfigs(res["indexConfigs"].([]interface{}))); err != nil {
```

File: mmv1/third_party/terraform/services/logging/resource_logging_project_bucket_config.go [hunk 14]
```diff
@@ -282,6 +309,10 @@ func resourceLoggingProjectBucketConfigRead(d *schema.ResourceData, meta interfa
 		return fmt.Errorf("Error setting cmek_settings: %s", err)
 	}
 
+	if err := d.Set("index_configs", flattenIndexConfigs(indexConfigs.([]interface{}))); err != nil {
```
Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_project_bucket_config.go:312]:
```suggestion
	if err := d.Set("index_configs", flattenIndexConfigs(res["indexConfigs"].([]interface{}))); err != nil {
```

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go (hunk 13 above)
Comment (rileykarson) [reply mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go:317]: (I haven't run these at all- you may want to make the changes locally and ensure it builds / a test passes to potentially save a round trip!)

File: mmv1/third_party/terraform/services/logging/resource_logging_project_bucket_config.go (hunk 14 above)
Comment (mudim) [reply mmv1/third_party/terraform/services/logging/resource_logging_project_bucket_config.go]: My bad, was rushing to get this change in. Resolved now

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go (hunk 13 above)
Comment (mudim) [reply mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go]: Yeah, have been for all builds. Should've paid close attn to latest one as it was the cache being on that tripped this over. Should've looked close to see the test results were cached for above commit

**Comment by mudim** [comment]
`make test` at provider level is now successful with no errors reported.
ATs running fine with below output

command: 
`make testacc TEST=./google/services/logging TESTARGS='-run=TestAccLoggingBucketConfigProject_indexConfigs'`

result:
```TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google/services/logging -v -run=TestAccLoggingBucketConfigProject_indexConfigs -timeout 240m -ldflags="-X=http://github.com/hashicorp/terraform-provider-google/version.ProviderVersion=acc"
=== RUN  TestAccLoggingBucketConfigProject_indexConfigs
=== PAUSE TestAccLoggingBucketConfigProject_indexConfigs
=== CONT TestAccLoggingBucketConfigProject_indexConfigs
--- PASS: TestAccLoggingBucketConfigProject_indexConfigs (43.29s)
PASS
ok 	http://github.com/hashicorp/terraform-provider-google/google/services/logging	47.747s

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go (hunk 13 above)
Comment (rileykarson) [reply mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go]: Oh, you're all good! I've just had a handful of cases like this turn into multiple day-long round trips of minor syntax changes enough times that I try to warn that I am suggesting a literal fix to a single syntax problem and not guaranteeing the code actually compiles or anything.

**Comment by rileykarson** [comment]
That doesn't seem right... let me try again.

/gcbrun

**Comment by mahesh-hmrc** [comment]
Hi @rileykarson, hope all well. Just wanted to follow up on above and if there is anything I can/need to do on my side. Thanks

**Comment by mudim** [comment]
Hi @rileykarson, hope all well. Just wanted to follow up on above and if there is anything I can/need to do on my side. Thanks

**Comment by rileykarson** [comment]
I'm gonna try this one more time before investigating deeper, sorry for the delay!

/gcbrun

File: mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go [hunk 15]
```diff
@@ -287,6 +314,10 @@ func resourceLoggingBucketConfigRead(d *schema.ResourceData, meta interface{}) e
 		return fmt.Errorf("Error setting cmek_settings: %s", err)
 	}
 
+	if err := d.Set("index_configs", flattenIndexConfigs(res["indexConfigs"].([]interface{}))); err != nil {
```
Comment (trodge) [inline mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go:317]: I suggest breaking this line up and first checking that `res["indexConfigs"]` is, in fact, a slice of `any` (or `interface{}`).

Comment (rileykarson) [inline mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go:317]: Ah, we're dying on this field when the value is `nil`. We'll want to change this to just `d.Set("index_configs", flattenIndexConfigs(res["indexConfigs"]));` and handle the value inside the flatten function. https://github.com/hashicorp/terraform-provider-google-beta/blob/35d78ff10768b60b27d020d08fde466d43826470/google-beta/services/compute/resource_compute_subnetwork.go#L1124-L1142 is a good reference.

Comment (rileykarson) [reply mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go:317]: https://github.com/GoogleCloudPlatform/magic-modules/pull/9109#discussion_r1367251751 instead- the stanza we use in generated code like subnetwork is resilient to as many edge cases as possible.

Comment (mudim) [reply mmv1/third_party/terraform/services/logging/resource_logging_bucket_config.go:317]: I see. Thanks for taking time with it, will get that updated now 👍

**Comment by mudim** [comment]
Hi @rileykarson, have pushed up change to handle null index_configs during crud. Thanks

**Comment by mudim** [comment]
Hi @rileykarson, hope all well. Just wanted to follow up on above and when you have few mins pls. Thanks

**Comment by rileykarson** [comment]
Running generate+test! Apologies, had some other work superseding most of my time for reviews.

**Comment by mudim** [comment]
Hi @rileykarson, just wanted to follow up the last set of VCR test results. From what I remember one of them was linked to an [issue](https://github.com/hashicorp/terraform-provider-google/issues/16099) you've raised already. Are we blocked on that for the approval on this one? Thanks

**Comment by trodge** [comment]
> Hi @rileykarson, just wanted to follow up the last set of VCR test results. From what I remember one of them was linked to an [issue](https://github.com/hashicorp/terraform-provider-google/issues/16099) you've raised already. Are we blocked on that for the approval on this one? Thanks

It shouldn't be blocking. The failing test is unrelated to this change, which is tested by `TestAccLoggingBucketConfigOrganization_indexConfigs`.

**Review by trodge — APPROVED** [review]

**Comment by ScottSuarez** [comment]
This change appears to have broken tests in the GA and BETA projects. The API is returning default fields where none were set by the user. @rileykarson shall we revert?

```
=== CONT  TestAccLoggingBucketConfigOrganization_basic
    vcr_utils.go:152: Step 1/4 error: After applying this test step and performing a `terraform refresh`, the plan was not empty.
        stdout
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        Terraform will perform the following actions:
          # google_logging_organization_bucket_config.basic will be updated in-place
          ~ resource "google_logging_organization_bucket_config" "basic" {
                id              = "organizations/*******/locations/global/buckets/_Default"
                name            = "organizations/*******/locations/global/buckets/_Default"
                # (6 unchanged attributes hidden)
              - index_configs {
                  - field_path = "jsonPayload.request.url" -> null
                  - type       = "INDEX_TYPE_STRING" -> null
                }
              - index_configs {
                  - field_path = "jsonPayload.response.status" -> null
                  - type       = "INDEX_TYPE_INTEGER" -> null
                }
            }
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccLoggingBucketConfigOrganization_basic (13.55s)
FAIL

```

Our tests are failing in these locations 
https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_Google_NightlyTests_GOOGLE_PACKAGE_LOGGING/54819?hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true&expandBuildTestsSection=true&expandBuildDeploymentsSection=false

**Comment by ScottSuarez** [comment]
Revert here https://github.com/GoogleCloudPlatform/magic-modules/pull/9448

**Comment by ScottSuarez** [comment]
Not reverting after talking with the team!
